src/main/java/edu/berkeley/cs186/database/index/LeafNode.java

```diff
@@ -146,42 +147,88 @@ private LeafNode(BPlusTreeMetadata metadata, BufferManager bufferManager, Page p
     // See BPlusNode.get.
     @Override
     public LeafNode get(DataBox key) {
-        // TODO(proj2): implement
-
-        return null;
+        return this;
     }
 
     // See BPlusNode.getLeftmostLeaf.
     @Override
     public LeafNode getLeftmostLeaf() {
-        // TODO(proj2): implement
 
-        return null;
+        return this;
     }
 
     // See BPlusNode.put.
     @Override
     public Optional<Pair<DataBox, Long>> put(DataBox key, RecordId rid) {
-        // TODO(proj2): implement
 
-        return Optional.empty();
+        if (keys.contains(key)) throw new BPlusTreeException("Not Allow Duplicate Key");
+
+        int d = metadata.getOrder();
+        int insertIdx = InnerNode.numLessThan(key, keys);
+
+        keys.add(insertIdx, key);
+        rids.add(insertIdx, rid);
+
+        Optional<Pair<DataBox, Long>> result = Optional.empty();
+        if (this.keys.size() == 2 * d + 1) {
+            LeafNode right = new LeafNode(metadata, bufferManager, keys.subList(d, keys.size()), rids.subList(d, rids.size()), rightSibling, treeContext);
+
+            DataBox split = keys.get(d);
+
+            keys = keys.subList(0, d);
+            rids = rids.subList(0, d);
+            rightSibling = Optional.of(right.getPage().getPageNum());
+
+            result = Optional.of(new Pair<>(split, right.getPage().getPageNum()));
+        }
+
+        sync();
+        return result;
     }
 
     // See BPlusNode.bulkLoad.
     @Override
     public Optional<Pair<DataBox, Long>> bulkLoad(Iterator<Pair<DataBox, RecordId>> data,
             float fillFactor) {
-        // TODO(proj2): implement
+        int maxSize = (int) Math.ceil(2 * metadata.getOrder() * fillFactor);
 
-        return Optional.empty();
+        Optional<Pair<DataBox, Long>> result = Optional.empty();
+
+        while (data.hasNext()) {
+            Pair<DataBox, RecordId> nextData = data.next();
+            keys.add(nextData.getFirst());
+            rids.add(nextData.getSecond());
+            if (keys.size() > maxSize) {
```

Comment:
```suggestion
            if (keys.size() == maxSize + 1) {
```
keys에는 하나씩 추가되니까 maxSize + 1보다 커질 경우는 없을 것 같습니다!